logs_sdk/lib/opentelemetry/sdk/logs/logger_provider.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Logs
+      # {LoggerProvider} is the SDK implementation of
+      # {OpenTelemetry::Logs::LoggerProvider}.
+      class LoggerProvider < OpenTelemetry::Logs::LoggerProvider
+        attr_reader :resource
+
+        EMPTY_NAME_ERROR = 'LoggerProvider#logger called without '\
+            'providing a logger name.'
+        FORCE_FLUSH_ERROR = 'unexpected error in OpenTelemetry::SDK::Logs::LoggerProvider#force_flush'
+
+        # Returns a new {LoggerProvider} instance.
+        #
+        # @param [optional Resource] resource The resource to associate with
+        #   new LogRecords created by Loggers created by this LoggerProvider.
+        # @param [optional Array] log_record_processors Log Record Processors to
+        #   associate with the LoggerProvider.
+        #
+        # @return [LoggerProvider]
+        def initialize(
+          resource: OpenTelemetry::SDK::Resources::Resource.create,
+          log_record_processors: []
+        )
+          @log_record_processors = log_record_processors
+          @mutex = Mutex.new
+          @resource = resource
+          @stopped = false
+        end
+
+        # Returns a {OpenTelemetry::SDK::Logs::Logger} instance.
+        #
+        # @param [optional String] name Instrumentation package name
+        # @param [optional String] version Instrumentation package version
+        #
+        # @return [OpenTelemetry::SDK::Logs::Logger]
+        def logger(name = nil, version = nil)
+          name ||= ''
+          version ||= ''
+
+          OpenTelemetry.logger.warn(EMPTY_NAME_ERROR) if name.empty?
+
+          @mutex.synchronize do
+            OpenTelemetry::SDK::Logs::Logger.new(name, version, self)
+          end
+        end
+
+        # Adds a new LogRecordProcessor to this {LoggerProvider}'s
+        # log_record_processors.
+        #
+        # @param [LogRecordProcessor] log_record_processor The new
+        #   LogRecordProcessor to add.
+        def add_log_record_processor(log_record_processor)
+          @mutex.synchronize do
+            @log_record_processors = @log_record_processors.dup.push(log_record_processor)
```

Comment:
This is copied from `TracerProvider#add_span_processor`. @fallwith and I were chatting about this LOC. He was asking about the use of `.dup.push` here.

Does `dup.push` provide atomicity in a way that the `push` method alone couldn't deliver? Or is there a different reason `dup.push` was used?